### ISSUE TITLE
Better error handling for database type mismatches

### ISSUE THREAD

**Issue by lukehutch** [serverpod/serverpod#1522]
**Describe the bug**

If I change something in the database schema, often the first indication I get of this is the following type of exception:

![image](https://github.com/serverpod/serverpod/assets/811305/8fb352af-1930-4f46-b19a-2255d0e7c786)

How is this even possible? It would seem the intent is for `_isNullableType` to only return true if `null` is a valid value for that type.

Can some more useful information be displayed in this case?

**Comment by lukehutch** [comment]
Actually `t` in this case is `bool`, so `_isNullableType` seems to be broken.

![image](https://github.com/serverpod/serverpod/assets/811305/3b55ab2a-e004-457e-87ef-be5267456f64)

**Comment by SandPod** [comment]
I agree that the error message is a bit vague. We should improve it so that it is clear that the data deserialized doesn't match the target data type.

That is the issue encountered in the first image. I would guess that you have changed a column (or field). from `bool?` to `bool` but you are still fetching `null` values from the database. This could happen if you have re-generated the project but not migrated the database.

The `isNullableType` method is actually working as expected, although the name should be improved. It's purpose is to inform us if the value that is being deserialized is either a nullable or a non nullable version the passed in type.

On the row in the second image it recognizes that `t`(that is passed from generated code) is either `bool` or `bool?`.

**Comment by lukehutch** [comment]
No, I think there's still a problem here.

I think

```dart
  bool _isType<T1, T2>(Type t) => t == T1 || t == T2;

  bool _isNullableType<T>(Type t) => _isType<T, T?>(t);
```

should be 

```dart
bool _isType<T>(Type t) => t == T;

bool _isNullableType<T>(Type t) => _isType<T?>(t);
```

because the test `t == T1 || t == T2` will always return `true` for `_isType<T, T?>(t)`.

The way to fix the larger issue though is something like the following, in generated protocol files, in `fromJson` methods, for fields that are not nullable:

Replace

```dart
      blocked:
          serializationManager.deserialize<bool>(jsonSerialization['blocked']),
```

with

```dart
      blocked:
          serializationManager.deserialize<bool>(jsonSerialization['blocked']
                  ?? throw 'Non-nullable field `blocked` is missing from JSON for `UserInfo`'),
```

So, to recap, I think there are two separate issues that need fixing here.

**Comment by lukehutch** [comment]
Addressed by #1527 and #1528.
